internal/serialization/schema.go

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization
+
+import (
+	"fmt"
+	"sort"
+)
+
+type Schema struct {
+	fieldDefinitionMap    map[string]*FieldDescriptor
+	typeName              string
+	// Go does not have TreeMap, so we use a slice to store sorted fields
+	fieldDefinitions      []*FieldDescriptor
+	id                    int64
+	numberOfVarSizeFields int32
+	fixedSizeFieldsLength int32
+}
+
+func NewSchema(typeName string, fieldDefinitionMap map[string]*FieldDescriptor, rabin RabinFingerPrint) Schema {
+	fieldDefinitions := make([]*FieldDescriptor, len(fieldDefinitionMap))
+	c := 0
+	for _, fd := range fieldDefinitionMap {
+		fieldDefinitions[c] = fd
+		c += 1
+	}
+	// Sort according to field name
+	sort.SliceStable(fieldDefinitions, func(i, j int) bool {
+		return fieldDefinitions[i].fieldName < fieldDefinitions[j].fieldName
+	})
+	schema := Schema{
+		typeName:           typeName,
+		fieldDefinitionMap: fieldDefinitionMap,
+		fieldDefinitions:   fieldDefinitions,
+	}
+	schema.init(rabin)
+	return schema
+}
+
+func (s *Schema) GetField(fieldName string) *FieldDescriptor {
```

Comment:
I think we should try to unify method receiver types for `Schema` as much as possible. We can make all of them value type except `init` or make all of them pointer type. Wdyt?